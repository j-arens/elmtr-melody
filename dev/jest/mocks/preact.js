module.exports = {
    unmountComponent: jest.fn(),
};
