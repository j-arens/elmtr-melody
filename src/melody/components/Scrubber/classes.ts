import { MELODY_CUSTOMIZATION_CLASS_PREFIX as c_prefix } from '@constants';
import { prefixClasses } from '@utils/index';

export const body = prefixClasses(
    c_prefix,
    'scrubber-body-bg-color',
);

export const backfill = prefixClasses(
    c_prefix,
    'scrubber-backfill-bg-color',
);

export const handle = prefixClasses(
    c_prefix,
    'scrubber-handle-bg-color',
);
