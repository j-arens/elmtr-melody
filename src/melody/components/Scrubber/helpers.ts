export const normalizeOffset = (offset: number): number => {
    if (offset > 100) {
        return 100;
    }
    if (offset < 0) {
        return 0;
    }
    return offset;
};

export const getNewTime = (offset: number, duration: number): number =>
    (normalizeOffset(offset) * duration) / 100;
