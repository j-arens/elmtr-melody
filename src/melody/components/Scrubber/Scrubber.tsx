import Slider from '@components/Slider/';
import { SliderBodyClickEvent, SliderDragEvent } from '@components/Slider/type';
import { Action } from '@redux/type';
import { MachineStates } from '@state-machine/type';
import { cySelector, NO_OP } from '@utils/index';
import { Component, h } from 'preact';
import * as classes from './classes';
import { getNewTime } from './helpers';

export interface StateProps {
    currentState: MachineStates;
    duration: number;
    currentTime: number;
}

export interface DispatchProps {
    updateCurrentTime: (updatedTime: number) => Action;
    toggleComponentDragging: (component: string, isDragging: boolean) => Action;
    triggerTimeSync: () => Action;
}

interface State {
    offset: number;
    dragging: boolean;
}

type Props = StateProps & DispatchProps;

export default class extends Component<Props, State> {
    state = {
        offset: 0,
        dragging: false,
    };

    componentDidUpdate(prevProps: Props, prevState: State) {
        const timeChanged = this.props.currentTime !== prevProps.currentTime;
        if (timeChanged) {
            const { duration, currentTime } = this.props;
            if (duration) {
                const nextOffset = 100 / (duration / currentTime);
                this.setState({ offset: nextOffset });
            } else {
                this.setState({ offset: 0 });
            }
        }

        const { dragging } = this.state;
        if (prevState.dragging !== dragging) {
            const { toggleComponentDragging } = this.props;
            toggleComponentDragging('scrubber', dragging);
        }
    }

    eventGuard = fn => (...args) => {
        if (!this.props.duration) {
            return NO_OP;
        }
        return fn(...args);
    }

    handleClick = this.eventGuard(({ insideHandle, offset }: SliderBodyClickEvent) => {
        if (insideHandle) {
            return;
        }
        const { updateCurrentTime, triggerTimeSync, duration } = this.props;
        updateCurrentTime(getNewTime(offset, duration));
        triggerTimeSync();
    });

    handleDragging = this.eventGuard(({ offset }: SliderDragEvent) => {
        const { updateCurrentTime, duration } = this.props;
        this.setState({ dragging: true });
        updateCurrentTime(getNewTime(offset, duration));
    });

    handleDragEnd = this.eventGuard(({ offset }: SliderDragEvent) => {
        const { updateCurrentTime, triggerTimeSync, duration } = this.props;
        this.setState({ dragging: false });
        updateCurrentTime(getNewTime(offset, duration));
        triggerTimeSync();
    });

    render(_, { offset }: State) {
        return (
            <Slider
                height={5}
                handleSize={14}
                eventRadius={8}
                throttle={10}
                offset={offset}
                onBodyClick={this.handleClick}
                onDragging={this.handleDragging}
                onEndDrag={this.handleDragEnd}
                classes={{
                    body: classes.body,
                    backfill: classes.backfill,
                    handle: classes.handle,
                }}
            />
        );
    }
}
