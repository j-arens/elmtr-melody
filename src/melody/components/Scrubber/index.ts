import { triggerTimeSync, updateCurrentTime } from '@redux/modules/audio/actions';
import { toggleComponentDragging } from '@redux/modules/ui/actions';
import { State } from '@redux/type';
import { connect } from 'preact-redux';
import { bindActionCreators, Dispatch } from 'redux';
import Scrubber, { DispatchProps, StateProps } from './Scrubber';

const mapState = (state: State) => ({
    currentState: state.machine.currentState,
    currentTime: state.audio.currentTime,
    duration: state.audio.filelength,
});

const mapDispatch = (dispatch) => bindActionCreators({
    updateCurrentTime,
    toggleComponentDragging,
    triggerTimeSync,
}, dispatch);

export default connect<
    StateProps,
    DispatchProps,
    {}
>(mapState, mapDispatch)(Scrubber);
