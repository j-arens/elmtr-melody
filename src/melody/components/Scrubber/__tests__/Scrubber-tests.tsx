import { NO_OP } from '@utils';
import { h } from 'preact';
import { render } from 'preact-render-to-string';
import Scrubber from '../Scrubber';

describe('Glider', () => {
    it('renders', () => {
        const component = render(
            <Scrubber
                currentState="stopped"
                duration={100}
                currentTime={0}
                updateCurrentTime={NO_OP}
                toggleComponentDragging={NO_OP}
                triggerTimeSync={NO_OP}
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
