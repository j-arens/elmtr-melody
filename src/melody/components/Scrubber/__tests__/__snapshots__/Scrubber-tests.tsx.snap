// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glider renders 1`] = `
<div class="slider elementor-tab-title undefined" style="padding: 8px 0;">
  <div class="slider__body melody-c-scrubber-body-bg-color" style="height: 5px;">
    <div class="slider__backfill melody-c-scrubber-backfill-bg-color" style="transform: translate3d(0%, 0, 0);"></div>
  </div>
</div>
`;
